### PR TITLE
Add a requirement to discard invalid ciphertexts

### PR DESCRIPTION
**Pull request by bifurcation**
This PR adds text requiring ciphertexts that do not decrypt to be discarded.  It also adds some discussion of the impact of this requirement on the safety of short tags.  (Thanks to @bren2010 for helping refine this analysis.)

@chris-wood - in #154 you mention addressing AEAD limits.  Could you suggest some text for that?  I was not able to interpret the CFRG draft.  Also, it seems like we would need some analogous analysis for the AES-CTR-based AEAD modes described here.  Do you have thoughts on that?

### CODE REVIEW

File: draft-ietf-sframe-enc.md [hunk 1]
```diff
@@ -521,7 +523,8 @@ def decrypt(metadata, sframe_ciphertext):
 
 If a ciphertext fails to decrypt because there is no key available for the KID
 in the SFrame header, the client MAY buffer the ciphertext and retry decryption
-once a key with that KID is received.
+once a key with that KID is received.  If a ciphertext fails to decrypt for any
+other reason, the client MUST discard the ciphertext.
```
Comment (Bren2010) [inline draft-ietf-sframe-enc.md:527]:
```suggestion
other reason, the client MUST discard the ciphertext. Invalid ciphertexts SHOULD be
discarded in a way that is indistinguishable (to an external observer) from having 
processed a valid ciphertext.
```

File: draft-ietf-sframe-enc.md [hunk 2]
```diff
@@ -873,6 +876,64 @@ The handling of replay is out of the scope of this document. However, senders
...
+* Media applications typically do not provide feedback to media senders as to
+  which media packets failed to decrypt.  When media quality feedback
+  mechanisms are used, decryption failures will typically appear as packet
+  losses, but only at an aggregate level.
+
+* Anti-replay mechanisms (see {{replay}}) prevent the attacker from re-using
+  legitimate ciphertexts.  Since the attacker is limited in their ability to
+  synthesize new legitimate (ciphertexts, tag) pairs (e.g., limited to flipping
+  bits in AES-CTR or AES-GCM ciphertexts), they can basically only cause the
+  media decoder to decode random data.
```
Comment (Bren2010) [inline draft-ietf-sframe-enc.md:923]:
```suggestion
```
Anti-replay doesn't protect other members of a call. I'm also not sure that the "can only decode random data" holds, given that plaintexts may be predictable and AES-CTR is malleable.

File: draft-ietf-sframe-enc.md [hunk 3]
```diff
@@ -873,6 +876,64 @@ The handling of replay is out of the scope of this document. However, senders
...
+
+In summary, with proper mitigations, short tags allow an attacker to send random
+data to a media stream at a very low rate, using an attack that is
+straightforward to recognize and neutralize.  At best, the attacker can trigger
+a denial-of-service condition, which they likely could have done anyway simply
+by sending meaningless packets, without regard for whether they pass the
+authentication check.
+
+Nonetheless, applications that make use of short tags need to put in place the
+mitigations discussed above.  In many cases, it is simpler to use full-size tags
```
Comment (Bren2010) [inline draft-ietf-sframe-enc.md:933]: Link to mitigations?

File: draft-ietf-sframe-enc.md [hunk 4]
```diff
@@ -873,6 +876,64 @@ The handling of replay is out of the scope of this document. However, senders
...
+  synthesize new legitimate (ciphertexts, tag) pairs (e.g., limited to flipping
+  bits in AES-CTR or AES-GCM ciphertexts), they can basically only cause the
+  media decoder to decode random data.
+
+In summary, with proper mitigations, short tags allow an attacker to send random
+data to a media stream at a very low rate, using an attack that is
+straightforward to recognize and neutralize.  At best, the attacker can trigger
+a denial-of-service condition, which they likely could have done anyway simply
+by sending meaningless packets, without regard for whether they pass the
+authentication check.
```
Comment (Bren2010) [inline draft-ietf-sframe-enc.md:930]: Would remove this paragraph as well

File: draft-ietf-sframe-enc.md (hunk 3 above)
Comment (bifurcation) [reply draft-ietf-sframe-enc.md:933]: The mitigations are the bullet points above.  There's not really a way to link to that.  But with the intervening paragraph removed, the reference should be clearer.

File: draft-ietf-sframe-enc.md (hunk 2 above)
Comment (bifurcation) [reply draft-ietf-sframe-enc.md:923]: All of these mitigations are receiver-local, in the sense that they're applied locally.  But it's a fair point about predictable plaintext, let me try to rephrase.

File: draft-ietf-sframe-enc.md [hunk 5]
```diff
@@ -873,6 +878,62 @@ The handling of replay is out of the scope of this document. However, senders
...
+  mechanisms are used, decryption failures will typically appear as packet
+  losses, but only at an aggregate level.
+
+* Anti-replay mechanisms (see {{replay}}) prevent the attacker from re-using
+  valid ciphertexts (either observed or guessed by the attacker).  A receiver
+  applying anti-replay controls will only accept one valid plaintext per CTR
+  value.  Since the CTR value is covered by SFrame authentication, an attacker
+  has to do a fresh search for a valid tag for every forged ciphertext, even if
+  the encrypted content is unchanged.  In other words, when the above brute
+  force attacke succeeds, it only allows the attacker to send a single SFrame
```
Comment (Bren2010) [inline draft-ietf-sframe-enc.md:927]:
```suggestion
  force attack succeeds, it only allows the attacker to send a single SFrame
```

File: draft-ietf-sframe-enc.md [hunk 6]
```diff
@@ -873,6 +878,62 @@ The handling of replay is out of the scope of this document. However, senders
...
+attacker_success_rate = attempts_per_second / 2^(8*Nt)
+```
+
+For example, a gigabit ethernet connection is able to transmit roughly 2^20
+packets per second.  If an attacker saturated such a link with guesses against a
+32-bit authentication tag (`Nt=4`), then the attacker would succeed on average
+roughly once every 2^12 seconds, or about once an hour.
+
+In a typical SFrame usage in a real-time media application, there are a few
+factors that mitigate this risk:
```
Comment (Bren2010) [inline draft-ietf-sframe-enc.md:903]:
```suggestion
approaches to mitigate this risk:
```

File: draft-ietf-sframe-enc.md [hunk 7]
```diff
@@ -873,6 +878,62 @@ The handling of replay is out of the scope of this document. However, senders
...
+  applying anti-replay controls will only accept one valid plaintext per CTR
+  value.  Since the CTR value is covered by SFrame authentication, an attacker
+  has to do a fresh search for a valid tag for every forged ciphertext, even if
+  the encrypted content is unchanged.  In other words, when the above brute
+  force attacke succeeds, it only allows the attacker to send a single SFrame
+  ciphertext; the ciphertext cannot be reused because either it will have the
+  same CTR value and be discarded as a replay, or else it will have a different
+  CTR value its tag will no longer be valid.
+
+Nonetheless, applications that make use of short tags need to put these
```
Comment (Bren2010) [inline draft-ietf-sframe-enc.md:932]:
```suggestion
Applications that make use of short tags should consider putting some or all of these
```

**Review by Bren2010 — APPROVED** [review]
nits

**Pull request merged**